Synthesize Equatable/Hashable for complex enums, structs

allevato opened the pull request:
<!-- What's in this pull request? -->

This is the implementation of the yet-to-be-merged/reviewed evolution proposal to synthesize `Equatable`/`Hashable` for enums with payloads and for structs (apple/swift-evolution#706). I'm offering this in the hopes that we can get the proposal reviewed, any updates if needed can be made to the implementation, and it can make it into Swift 4.

This is my first foray into a big change to the compiler itself, so I'm sure there are a number of subtleties about building these ASTs that may have gone over my head. I'm very open to any feedback anyone has on these changes!

allevato committed:
[AST] Accessors for ^= and _mixInt

allevato committed:
[Sema] Synthesize Eq/Hash for enums with payloads, structs

allevato committed:
[AST] Add identifier for synthesized struct equals

allevato committed:
[AST] Enable derived Equatable/Hashable for structs

allevato committed:
[AST] Add DeclContext.getAsStructOrStructExtensionContext

allevato committed:
[test] Update enum Equatable/Hashable tests

CodaFi commented on lib/AST/ASTContext.cpp:
```diff
@@ -908,6 +914,75 @@ FuncDecl *ASTContext::getEqualIntDecl() const {
+
+  auto intType = getIntDecl()->getDeclaredType();
+  auto inoutIntType = InOutType::get(intType);
+  SmallVector<ValueDecl *, 30> xorFuncs;
```
Nit: Are there actually 30 of these?  Usually SmallVector sizes are powers of 2, so 32 would be more apropos

allevato replied:
Thanks for pointing that out—I didn't go back and change it after I cribbed the function from `==`. It looks like there are currently ~12, so I'll use 16.

CodaFi commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -172,44 +371,228 @@ static void deriveBodyEquatable_enum_eq(AbstractFunctionDecl *eqDecl) {
+  
   BraceStmt *body = BraceStmt::create(C, SourceLoc(), statements, SourceLoc());
   eqDecl->setBody(body);
 }
```
Please check your editor's whitespace setting and remove the above space changes.

allevato replied:
Done.

CodaFi commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -32,26 +32,181 @@
+    } else {
+      // One associated value with no label is represented as a paren type.
+      auto actualType = argumentType->getWithoutParens();
+      if (!typeConformsToProtocol(tc, declContext, actualType, protocol)) {
```
Nit: This, and the call above it, can be used as the return value of this branch.

allevato replied:
How so? Unless I'm misunderstanding you, we only want to early-exit if we find a non-conforming type; we have to see the entire loop through before we can know that we can return true.

CodaFi commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -32,26 +32,181 @@
+    theStruct->getStoredProperties(/*skipInaccessible=*/true);
+  for (auto propertyDecl : storedProperties) {
+    auto propertyType = propertyDecl->getType();
+    if (!typeConformsToProtocol(tc, declContext, propertyType, protocol)) {
```
Same here

allevato replied:
Same comment.

CodaFi commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -318,13 +701,38 @@ ValueDecl *DerivedConformance::deriveEquatable(TypeChecker &tc,
-      return deriveEquatable_enum_eq(tc, parentDecl, theEnum);
+    if (auto theEnum = dyn_cast<EnumDecl>(type)) {
+      auto bodySynthesizer = theEnum->hasOnlyCasesWithoutAssociatedValues()
+      ? &deriveBodyEquatable_enum_noAssociatedValues_eq
```
Might want to run clang-format in this region.

allevato replied:
Done.

CodaFi commented:
Overall it looks pretty good.  I'll have more in-depth comments later.

CodaFi commented on test/Sema/struct_equatable_hashable.swift:
```diff
@@ -0,0 +1,144 @@
+
+// FIXME: This should work.
+func overloadFromOtherFile() -> YetAnotherFromOtherFile { return YetAnotherFromOtherFile(v: 1.2) }
+func overloadFromOtherFile() -> Bool { return false }
```
What's up here?

allevato replied:
Oops, forgot to go back and remove that after pulling the examples over from the `enum` version of the tests. (I'm guessing the bug that was the reason for the `FIXME` in the `enum` version was fixed but the comment mistakenly remained in that file?)

I've removed the `FIXME` from the other file as well.

CodaFi commented on lib/AST/ASTContext.cpp:
```diff
@@ -908,6 +914,75 @@ FuncDecl *ASTContext::getEqualIntDecl() const {
   return nullptr;
 }
 
+FuncDecl *ASTContext::getMutatingXorIntDecl() const {
```
Thinking out loud: This code has been copied and pasted enough times it's probably worth it to factor it out at some point...

allevato replied:
I'll tinker with factoring the commonalities out into a function that takes a callback and loop back around.

allevato replied:
I've just pushed a new commit that factors out the common code for this and for the intrinsic library function lookups.

CodaFi commented on lib/AST/Decl.cpp:
```diff
@@ -2099,8 +2099,14 @@ bool NominalTypeDecl::derivesProtocolConformance(ProtocolDecl *protocol) const {
+      // DerivedConformance::derive{Equatable,Hashable} perform the full check
+      // and it outputs its own diagnostics if conformance cannot be
+      // synthesized.
+      return enumDecl->hasCases();
```
If this check has to become semantic, maybe libAST isn't the right place for it.

CodaFi replied:
If you can, look into moving this into the type checker itself, and diagnosing the specific members that are hindering synthesis

allevato replied:
Since this is my first time really diving into the various compiler components, can you elaborate a little more on what you mean here w.r.t moving it into the type checker?

Re: diagnostics, I think it would be great addition to be able to emit diagnostics that say "*protocol* couldn't be synthesized because _such and such_ does not conform to *protocol*", but this is also something that we'd want across the board for `Codable` and future derived protocols as well. In the interest of scoping this PR, could that be something added to both later?

CodaFi replied:
It absolutely could.  The layering here makes me uncomfortable all the same, which is the first half of the comment.  

To resolve this: take a look at all the places this accessor is called, I'll bet you most/all of them are in Sema.  If that's the case, then move it down into the type checker itself and quash the FIXME.

allevato replied:
Understood! It looks like this is *only* called from Sema, and all those call sites provide easy access to `TypeChecker` so this should be fairly straightforward to refactor. I'll work on it.

allevato replied:
Just pushed a new commit. I didn't move the method into the type checker per se, but into the `DerivedConformance` namespace where it seemed more appropriate (and it now takes a TC as an argument, so it can do all the derivability checks at once). I resolved the FIXME for Equatable/Hashable but left Codable as-is for now. PTAL.

allevato committed:
Formatting and other cleanup.

allevato commented:
Thanks for the feedback! I've pushed the low-hanging formatting clean-up and am working on the more involved items.

allevato committed:
[AST] Factor out common code in operator/func lookups

allevato committed:
Move derivesProtocolConformances from AST to Sema

allevato committed:
[Sema] Forbid synthesis in an extension

The exception is if the conformance is already redundant (for example,
the implicit synthesis for RawRepresentable enums).

allevato commented:
I've updated this PR (and the proposal) to forbid synthesis in extensions, to stay in sync with the recently merged `Codable` change (#9758).

CodaFi commented:
@swift-ci please smoke test

allevato commented:
I'm a bit puzzled at why the `swiftpm` build is crashing on Linux but not on macOS in that smoke test (especially without a stack trace to look at).

I'll try to get a development environment set up on a Linux instance but it'll take a little while to carve out that time.

CodaFi commented:
@allevato Could you rebase?  I'll try to run this on a Linux box for you.

allevato commented:
@CodaFi I've rebased and pushed. Thanks for taking a look at this!

allevato commented:
I did a little more digging and found that the [smoke tests](https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#smoke-testing) on macOS don't build Swift Package Manager, which explains why the problem didn't manifest there. I built it explicitly on my Mac and got the same crash that was in the CI logs, so now I have something to debug with.

allevato committed:
[Sema] Fix type checking for enums defined in other files

allevato commented:
I've pushed a new commit that fixes the type checker issue, added unit tests to verify the fix, and also verified that Swift Package Manager builds with this change.

Can you please kick off another smoke test?

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

allevato committed:
[Sema] inout/ASTNode fixes after rebasing

CodaFi commented:
@swift-ci please smoke test

lattner commented on lib/AST/ASTContext.cpp:
```diff
@@ -178,6 +178,12 @@ FOR_KNOWN_FOUNDATION_TYPES(CACHE_FOUNDATION_DECL)
+  /// func ^= (inout Int, Int) -> Void
+  FuncDecl *MutatingXorIntDecl = nullptr;
+
+  /// func _mixInt(Int -> Int
```
typo, needs an extra )

allevato replied:
Done.

lattner commented:
Please add a test case for indirect enums.  Thanks!

allevato committed:
Fix typo in _mixInt comment

allevato committed:
[test] Add tests for indirect enums

natecook1000 commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -376,34 +1017,64 @@ deriveHashable_enum_hashValue(TypeChecker &tc, Decl *parentDecl,
+  //     var result: Int
+  //     switch self {
+  //     case A:
+  //       result = _mixInt(0)
```
I think we could devise a more efficient function to use here than `_mixInt`. Its purpose is to increase the variability of hash values before using them with hashed collections, so calling it multiple times does more work than is necessary. For example, each call incorporates the per-execution seed, which isn't necessary at the instance level if it's happening at the type level.

Additionally, an enum case with no associated values or a struct with only a single property shouldn't need mixing at all. That is, this could just be `result = 0`, while the struct implementation could start with `var result = x.hashValue`.

natecook1000 replied:
ps. This change is the best. Sorry I didn't lead with that! :clap::clap::clap::100:

allevato replied:
Thanks!

My original implementation actually used a multiplicative hash, but I switched over to `_mixInt` once I discovered that it existed and seeing it used in the `Set<T>` implementation here: https://github.com/apple/swift/blob/5d17b31948b0ff7132149e9e86f254e69328934e/stdlib/public/core/HashedCollections.swift.gyb#L1146

Interestingly, they also start by mixing zero, which I presumed was for ensuring that the distribution isn't biased by edge cases, which in the case of `Set` would be the empty set—that's why mix the ordinals here as well.

I'm by no means a hashing expert, so I'd appreciate more input from folks who are—implementing a new hash function in stdlib and swapping out for `_mixInt` for it in this code would be straightforward, but it would be good to have some consensus on what that hash function should look like before I start down the road of an implementation 🙂

natecook1000 replied:
That makes sense! There are lots of to-dos in that Hashing.swift file, so I know it's on the "when there's time for it" pile for the standard library. cc @airspeedswift 

Part of the `Set` implementation is that it has an unusual requirement—two sets can be equal even if their elements are in different orders, and equal sets must have the same hash value. Using XOR to merge mixed hash values solves that, since XOR is commutative/associative in the way that a multiplicative hash wouldn't be (I think).

DougGregor commented:
@swift-ci please smoke test

freak4pc commented on test/Sema/enum_equatable_hashable.swift:
```diff
@@ -89,6 +88,59 @@ enum Complex {
+var genericHashableHash: Int = GenericHashable<String>.A("a").hashValue
+
+// But it should be an error if the generic argument doesn't have the necessary
+// constrants to satisfy the conditions for derivation.
```
Not a super buggy but I was just reading through this PR and noticed `constrants` -> `constraints` :)

allevato replied:
Thank you for reading so closely! Fixed.

allevato committed:
Fix some typos

benrimmington commented:
@allevato Pull request #11735 might be of interest to you.

allevato commented:
@benrimmington Thanks for pointing me to that. Once that gets merged, I'll update this PR to mirror that behavior. (I presume it would be noncontroversial to do that here, for consistency.)

slavapestov commented on include/swift/AST/DeclContext.h:
```diff
@@ -255,6 +256,10 @@ class alignas(1 << DeclContextAlignInBits) DeclContext {
 
+  /// If this DeclContext is a struct, or an extension on a struct, return the
+  /// StructDecl, otherwise return null.
+  StructDecl *getAsStructOrStructExtensionContext() const;
```
You might want to do a pass over callers of getAsNominalTypeOrNominalTypeExtensionContext() and see if any just `dyn_cast<StructDecl>` the result and clean them up.

allevato replied:
I found 98 instances of `getAsNominalTypeOrNominalTypeExtensionContext`, but none of them appear to unconditionally cast to `StructDecl` afterward.

slavapestov commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -32,26 +32,186 @@
+/// \p type The type whose conformance should be checked.
+/// \p protocol The protocol being requested.
+/// \return True if the type conforms to the protocol.
+static bool typeConformsToProtocol(TypeChecker &tc, DeclContext *context,
```
I feel like this utility does not carry its weight. Note that an Optional value already has a conversion to bool, so you can just call 'if (tc.conformsToProtocol(...))', I think.

allevato replied:
Removed.

slavapestov commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -148,13 +352,13 @@ static void deriveBodyEquatable_enum_eq(AbstractFunctionDecl *eqDecl) {
   assert(cmpFunc && "should have a == for int as we already checked for it");
-  
+
   auto fnType = cast<FunctionType>(cmpFunc->getInterfaceType()
```
I don't think you need the getCanonicalType() calls in this function, instead do `cmpFunc->getInterfaceType()->castTo<FunctionType>()`

allevato replied:
Done.

slavapestov commented on lib/Sema/DerivedConformanceEquatableHashable.cpp:
```diff
@@ -185,33 +389,217 @@ static void deriveBodyEquatable_enum_eq(AbstractFunctionDecl *eqDecl) {
+  }
+
+  // switch (a, b) { <case statements> }
+  auto aRef = new (C) DeclRefExpr(ConcreteDeclRef(aParam), DeclNameLoc(),
```
The ConcreteDeclRef ctor is implicit so you don't have to call it explicitly (you rely on the implicit behavior below)

allevato replied:
Removed throughout.

slavapestov commented on lib/Sema/DerivedConformances.cpp:
```diff
@@ -23,7 +23,87 @@
+    // a type conforms manually).
+    if (*knownProtocol == KnownProtocolKind::Encodable ||
+        *knownProtocol == KnownProtocolKind::Decodable) {
+      // FIXME: This is not actually correct. We cannot promise to always
```
But you're passing in a type checker instance to this function, no?

allevato replied:
That's true now. This code was moved wholesale from `ASTContext` to `DerivedConformances` at the recommendation that it belonged in Sema. Since IIRC this is originally @itaiferber's code/comment relating to `Codable`, I didn't want to touch anything outside the scope of `Equatable`/`Hashable`.

slavapestov replied:
It's fine to clean up other code (and in fact encouraged) when implementing new features.

allevato replied:
Good to know for the future! If you're interested in merging this as-is (referencing your comment below), then I'm happy to take a look at that cleanup in another PR.

slavapestov commented:
@allevato Can you also add some execution tests? So far you're just ensuring the derived conformance type checks, but it might hit problems in SILGen, etc, or even at runtime.

allevato committed:
[Sema] Apply review feedback

- Remove unnecessary helper function
- Remove explicit ConcreteDeclRef ctor calls
- Clean up some function casts

allevato commented:
@slavapestov Thanks for the review pass! Regarding executable tests, `test/Sema/enum_equatable_hashable_correctness.swift` and `test/Sema/struct_equatable_hashable_correctness.swift` were included in 5f3965915517629d5d348596992ac52515917efb. Are they not working as intended?

allevato committed:
Merge branch 'master' into synthesize-equatable-hashable

natecook1000 commented:
@allevato Could you add tests that verify the semantics for `Hashable` synthesis? We don't want to test for individual hash values, since those can change, but the `hashValue` for two equal instances should be equal, and the `hashValue` for two nearly-but-not-equal instances should be *not* equal.

jrose-apple commented:
Note: this is what StdlibUnittest's `checkHashable` is for.

natecook1000 commented:
`checkHashable` checks the semantic requirement (equal instances must have equal hash values), but doesn't check whether non-equal instances have different hash values. That's probably okay for the tests in general, since collisions don't violate the semantics of `Hashable`, but for this generated conformance it'd be great to make sure we're at least getting the basics right.

`checkHashable` also verifies `Equatable` conformance, so the custom `==` tests could be dropped once that's being used.

allevato committed:
Improve synthesized member correctness tests

- Migrate them to use StdlibUnittest
- Add tests that "close" values do not have colliding hash values

allevato committed:
Merge branch 'master' into synthesize-equatable-hashable

allevato committed:
"yet" was removed from some diagnostics by 21b2073b

allevato commented:
@natecook1000 @jrose-apple I've rewritten the correctness tests to use `StdlibUnittest`, leverage `checkHashable` for the standard equatability/hashability checks, and added some "nearby values don't collide" checks.

I've been thinking more about the hash function issues we talked about above, and how the XOR operation in particular isn't appropriate here. Would it be reasonable to add a new mixing function to [Hashing.swift](https://github.com/apple/swift/blob/master/stdlib/public/core/Hashing.swift) specifically for these synthesized operations? Something like:

```swift
public func _mixForSynthesizedHashValue(_ accumulator: Int, _ value: Int) -> Int {
  // some implementation
}
```

Then refinements to that hash function are isolated there, without requiring low-level changes to the AST nodes synthesized by the compiler. We could start with a simple fast polynomial implementation (`return 31 &* accumulator &+ value`), or something else if you have any other recommendations.

What do you think?

natecook1000 commented:
@allevato The tests look great, thanks! 

I think it does makes sense to pull the hashing operation out into the stdlib, since it will be easier to improve there, if necessary. The biggest issue with XOR hashing is that it's a commutative operation, so values like `Point(x: 10, y: 20)` and `Point(x: 20, y: 10)` end up with the same hash value. That's desired in `Set`'s implementation, but not here. Your suggested replacement should be an improvement on that front—the hashed collections use `_mixInt` to improve the distribution of hash values as long as they're mostly unique.

allevato committed:
[stdlib] Add _mixForSynthesizedHashValue to stdlib

allevato committed:
Synthed hashValue now calls _mixForSynthesizedHashValue

For enums, we now only call _mixInt on the computed hash value if
it has associated values; for enums without associated values, we
use the ordinal alone, as before.

allevato committed:
[test] Fix executable_test requirement

allevato commented:
A handful of updates:

I've moved the hash function into Hashing.swift and call it from the synthesized impl. Right now it just computes a polynomial using `31 &* oldValue &+ nextValue` for each associated value or struct field. Then, the synthesized `hashValue` calls `_mixInt` at the end. This is possibly redundant since `Set` and `Dictionary` both end up calling `_mixInt` on the element type's hash value before bucketing, but I noticed that a number of other stdlib types also call `_mixInt` as part of their own `hashValue`.

For enums that don't have associated values, the latest commits revert `hashValue` back to just using the ordinal without calling `_mixInt` at all, since sequential ordinals are already optimally distributed.

I also realized that my new tests weren't actually running in the whole suite because I wrote `executable-test` instead of `executable_test`, so... that's fixed. 😥 

I've found one more problem that I'm not clear on how to fix—for indirect enums, whenever my synthesized `hashValue` accesses the `hashValue` of an associated value of the same type, I get the following warnings:

```
<unknown>:0: warning: attempting to access 'hashValue' within its own getter
<unknown>:0: note: access 'self' explicitly to silence this warning
```

This doesn't make sense, because I never access `self`'s hash value implicitly—I always access the hash value of an associated value expression. It feels like the [`UnresolvedDotExpr` here](https://github.com/apple/swift/pull/9619/files#diff-050bbcdd54b1180410b9421b59aa8137R758) is a culprit, but it's not obvious to me what's going wrong. Any ideas?

allevato committed:
[Sema] Only emit warnings for implicit self use in getters

This fixes a pre-existing bug where implicit DeclRefExprs involving
*any* expression of the same type within a getter were flagged with
warnings, even if they were references to something other than "self".

allevato commented:
Ah-ha, I found the problem: a bug in [MiscDiagnostics.cpp](https://github.com/apple/swift/blob/8b31e1e4427c5f98dcdfd02a4dc60507110e2b78/lib/Sema/MiscDiagnostics.cpp#L1532). That line is emitting diagnostics for *any* implicit `DeclRefExpr` involving the same type, not just the ones where the base expression is `self`. I've pushed a commit that fixes that and the errors for indirect enums go away.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented on test/Sema/struct_equatable_hashable.swift:
```diff
@@ -0,0 +1,154 @@
+  }
+  var hashValue: Int { return 0 }
+}
+// ...but synthesis in a type defined in another file doesn't work yet.
```
Can you file a bug to investigate this if you don't plan on doing it in this PR?

allevato replied:
Filed [SR-6101](https://bugs.swift.org/browse/SR-6101).

slavapestov commented:
@allevato @CodaFi @jrose-apple Can we merge this and deal with any follow-up work later? It would be good to get some wider testing here.

allevato committed:
[test] Move executable tests into test/Interpreter

slavapestov commented:
@allevato Can you rebase on top of master, and we can merge this in?

allevato committed:
Merge branch 'master' into synthesize-equatable-hashable

slavapestov commented:
@allevato That's fine then, let's leave the history as-is. Looks like there's one more merge conflict (sorry), once you fix that I'll merge this PR. Thanks!

allevato committed:
Merge branch 'master' into synthesize-equatable-hashable

allevato commented:
@slavapestov Done—this should be ready to merge now. Thanks! 😄

slavapestov commented:
Alright, thank you!

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test macOS

slavapestov commented:
Looks like CI is down for maintenance. Should be back soon

allevato commented:
It looks like that failure might be unrelated, but can someone else verify?

```
<unknown>:0: error: module map file '/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift-corelibs-libdispatch/dispatch/module.modulemap' not found
<unknown>:0: error: missing required module 'SwiftShims'
ninja: build stopped: subcommand failed.
```

DougGregor commented:
@swift-ci please clean smoke test Linux

DougGregor commented:
@swift-ci please clean smoke test Linux

DougGregor commented:
Yeah, it's unrelated. Would like to get actual Linux testing though

DougGregor commented:
@swift-ci please clean smoke test Linux

DougGregor commented:
@swift-ci please clean test Linux

therealbnut commented:
🎉 congratulations @allevato et al. I look forward to seeing this in an upcoming Swift release! (hopefully not too far away 😉)

lattner commented:
Great work @allevato !

solidcell commented:
Thanks so much @allevato! This was one of my top most wanted features in Swift.

Is this likely to make it into a specific upcoming version? I'm not sure how release coordination works for Swift.

lattner commented:
This is most likely to go out in "swift 4.1" in spring 2018

bielikb commented:
Hi guys, today I've tried snapshot of Swift 4.1. I've tried to define enum with associated values in which the value also conforms to Equatable. 

```
enum EnumWithAssociatedValues : Equatable {
    case first(String)
    case second(String)
}
```

However the conformance was not added and compiler threw couple of errors. Is it possible I misread the proposal and enums/value types with associated values are not yet supported/not part of this proposal? Thanks for your answer in advance.

xwu commented:
@bielikb Your example works without any errors for me. Are you set up to use the toolchain you downloaded? (Xcode Playgrounds, for example, do not.)

bielikb commented:
Ive tried to use both - terminal (set local version using swiftenv) & xcode playgrounds (set toolchain). Ill check it tomorrow. Equatable/Hashable conformance worked for enums/structs with no associated types without any problem.

ematejska commented:
@bielikb If you want to easily try it and make sure that you're using the correct toolchain do the following:
1. Launch Xcode and select Xcode->Toolchains-><your downloaded toolchain>
2. Create a new command line project for MacOS  (File->New->macOs->Command Line Tool)
3. Go to your project and select the main.swift file and paste your code you want to test.

The problem you have is that Playgrounds will not respect the toolchain setting so using those will not work for testing.

bielikb commented:
@ematejska Thanks for hints, that's something I already know. 
Im getting errors also for swift script file. 

Could you try to build the Swift script that I attached to this comment, if it works for you? Maybe then it's env related problem, rather then implementation itself. When I run `Enums.swift` with associated enums, I got couple of compiler errors. If I remove associated values from declaration, it works just fine.
Please note that the attached zip contains also `.swift-version` file that specifies 4.1 snapshot I'm using.

Feel free to correct me, if I'm doing something wrong.

[enum_associated_values.zip](https://github.com/apple/swift/files/1562249/enum_associated_values.zip)

ematejska commented:
@bielikb  I already had the 4.1-Snapshot-2017-12-13(a) toolchain installed so I tried it with that and Xcode 9.  Seems to work for me.  I saw errors with the default toolchain that was shipped with Xcode 9 or the form "error: type 'EnumWithAssociatedValues' does not conform to protocol 'Equatable'" and then when I switched to the 4.1-Snapshot, I was able to build successfully.

Just to be clear, I didn't run this as a script but pasted your code into my command line project as described in the previous comment.

bielikb commented:
@ematejska I've tried to run it in XCode 9.1 in sample project and it worked! As you mentioned. I however still get problems to build it as a script - but that might be problem of choosing the correct Swift toolchain - e.g. although I specified 4.1 snapshot as local swift version for dir with my Enums.swift, it's probably picking 4.0.2 toolchain. Thanks for help & getting me out of the "fog". I cannot wait until 4.1 will hit the world, good night boiler plate.

benrimmington commented:
@bielikb Try including the xctoolchain alias/identifier in your script:

```Shell
#!/usr/bin/xcrun --toolchain 'swift' swift
```

or

```Shell
#!/usr/bin/xcrun --toolchain 'org.swift.4120171212a' swift
```

bielikb commented:
@benrimmington I've tried the former definition and it worked! Thanks a lot guys ;)
